### DISCUSSION TITLE
Nethserver (Centos7) system service

### DISCUSSION THREAD

**Discussion by tdcockers** [abraunegg/onedrive#2024] [Q&A]
I'm having issues getting the client to autostart on a Nethserver instance, which is essentially a web interface overlay for a headless Centos 7 install. I'm almost certain my issues stem from me not understanding how linux authentication/privilege escalation work, so hopefully the answer is simple enough.

Client is installed, running, authorised and manually syncing happily (sharepoint, 4 x libraries/instances) using --confdir and --monitor, all from 'root' login to the server (ie no sudo).

When I try enable the services, using 'systemctl enable/start custom.service' fails and the log shows it asking for authorisation again (configs are under /root/.config/onedrive/sharepoint/{various}). I presume it is running in a different 'root' profile or something like that?

The server doesn't have any other user accounts on it. I tried a various assortment of ways to run it as a user service using the root account as well just in case, but keep getting errors along the lines of failed to connect to DBus, no such file.

Any tips? Thanks.

**Comment by abraunegg** [comment]
@tdcockers 
> When I try enable the services, using 'systemctl enable/start custom.service' fails and the log shows it asking for authorisation again (configs are under /root/.config/onedrive/sharepoint/{various}).

When the systemd service is running and generates that authentication prompt - what is the actual user running those service files? That 'user' may not actually have access to your configuration files under 'root'.

Generally - as a security aspect - you should not be running services as root ... but you are free to do as you want.

> I presume it is running in a different 'root' profile or something like that?

There is only 1 'root' login ... 'root'  is 'root'.

What is probably happening here is that *however* you are starting the services / enabling the services - and is probably something unique to your chosen distribution. It may be running | enabling the services as a different user.

> The server doesn't have any other user accounts on it. I tried a various assortment of ways to run it as a user service using the root account as well just in case, but keep getting errors along the lines of failed to connect to DBus, no such file.

This is curious .. if you are getting `Failed to connect to bus: No such file or directory` as an error message - this means that your Nethserver platform is trying to enable systemd services potentially in the following manner:
```
systemctl --user enable onedrive
systemctl --user start onedrive
```
But .. CentOS 7 does *not* support `--user` directives ... if can only be started as follows:
```
systemctl enable onedrive
systemctl start onedrive
```

See: https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#onedrive-service-running-as-root-user-via-systemd-red-hat-enterprise-linux-centos-linux

If Nethserver is trying to do something different in enabling systemd services - that means the systemd package provided by Nethserver is not the same as what is provided by CentOS.

I suspect a number of items here going on for you:
1. Make sure your modified systemd files are in the correct location for your platform
2. Make sure that your 'confdir' path for each configuration is correct, exists, and contains a copy of the 'config' file
3. Make sure that each of these systemd files have the correct `ExecStart` path: `ExecStart=/usr/local/bin/onedrive --monitor --confdir="/full/path/to/config/dir"` specifying the full path to the
4. If your platform *really is* CentOS 7 based, enable the service as per above:
```
systemctl enable custom-service-for-onedrive.service
systemctl start custom-service-for-onedrive.service
```
5. The service should start without issue
6. Check the systemd logs as per:
```
journalctl --unit=custom-service-for-onedrive.service -f
```

If this does not work - this is not an issue I can assist with any further - and you need to speak to the folk who provide support for your chosen distribution: https://community.nethserver.org/categories

**Comment by tdcockers** [comment]
Thank you, that is about what I had gotten to... just to be clear, I wasn't using the Nethserver interface for setting this up but rather using the CLI to going through the process that you've documented, and clarified above. NS does have the ability to control the systemd services but only if you explicitly set it up.

I think I've tried the suggestions above... off the the NS forums I suppose!
Thanks,

**Comment by abraunegg** [comment]
@tdcockers 
Another suggestion for you would be to change your systemd configuration to use `--verbose --verbose` and enable logging to a logfile via modifying your applicable 'config' file.

The combination of those changes will help identify what is going on in a systemd scenario to work out what the application is doing in terms of interpreting where the config file is located and *why* it cant find it.

If you can do this testing and provide the logs  this will go a long way to helping you resolve this issue.

**Comment by tdcockers** [comment]
Hey mate, thanks for checking back in.

I've run the service with the line:
``` ExecStart=/usr/local/bin/onedrive --monitor --confdir="/etc/.od-config/scans" --verbose --verbose > /etc/.od-config/scans/debug_output.log 2$>1```
nothing is generated; journalctl still shows the service asking for M365 authorisation.

Running the onedrive line with the debug output manually works as expected. Service file is in /lib/systemd/system/.

I guess there is some sort of permission issue going on, but I just can't see where it is. Haven't had any solutions offered from the NS forums that have helped yet either :(

**Comment by abraunegg** [comment]
@tdcockers 
Is anything at all written to that debug log?

What happens if you take all the redirect out, leave verbose, and in the 'config' file enable logging

The verbose systems logs should show what the application is calculating where to find the config

**Comment by tdcockers** [comment]
No, nothing at all written, and if the file doesn't exist before it isn't being created either.

I've removed the redirect, enabled logging in the config and set the log directory to /var/log/onedrive/scans. Starting the service logs nothing, starting manually via command line logs the following:
```
022-Jul-03 15:52:22.1972167    [DEBUG] Process ID: 7FD75A92B260
2022-Jul-03 15:52:22.197393     [DEBUG] User UID:   0
2022-Jul-03 15:52:22.1974174    [DEBUG] User GID:   0
2022-Jul-03 15:52:22.1974378    [DEBUG] User Name:  root
```

Could it possibly be a permission issue on the binary rather than the config/logging/share directories?

**Comment by abraunegg** [comment]
@tdcockers 
What is journalctl showing for the service .. there should be systemd logs of the debug output

**Comment by abraunegg** [comment]
@tdcockers 
After installing NetServer 7 myself, and installing the client following the directions for CentOS 7, then authenticating the client, then enabling the client via systemd - I can find *zero* issue ...
* https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md#dependencies-fedora--version-18--centos-7x--rhel-7x
* https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#onedrive-service-running-as-root-user-via-systemd-red-hat-enterprise-linux-centos-linux

**Compile, Install and Authenticate**
```
[root@nethserver-test ~]# curl -fsS https://dlang.org/install.sh | bash -s dmd-2.099.0
Downloading https://dlang.org/d-keyring.gpg
######################################################################## 100.0%
Downloading https://dlang.org/install.sh
######################################################################## 100.0%
gpg: directory `/root/.gnupg' created
gpg: new configuration file `/root/.gnupg/gpg.conf' created
gpg: WARNING: options in `/root/.gnupg/gpg.conf' are not yet active during this run
gpg: keyring `/root/.gnupg/pubring.gpg' created
gpg: /root/.gnupg/trustdb.gpg: trustdb created
The latest version of this script was installed as ~/dlang/install.sh.
It can be used it to install further D compilers.
Run `~/dlang/install.sh --help` for usage information.

Downloading and unpacking http://downloads.dlang.org/releases/2.x/2.099.0/dmd.2.099.0.linux.tar.xz
######################################################################## 100.0%
Using dub 1.28.0 shipped with dmd-2.099.0

Run `source ~/dlang/dmd-2.099.0/activate` in your shell to use dmd-2.099.0.
This will setup PATH, LIBRARY_PATH, LD_LIBRARY_PATH, DMD, DC, and PS1.
Run `deactivate` later on to restore your environment.
[root@nethserver-test ~]# source ~/dlang/dmd-2.099.0/activate
(dmd-2.099.0)[root@nethserver-test ~]# git clone https://github.com/abraunegg/onedrive.git
Cloning into 'onedrive'...
remote: Enumerating objects: 5871, done.
remote: Counting objects: 100% (202/202), done.
remote: Compressing objects: 100% (102/102), done.
remote: Total 5871 (delta 118), reused 163 (delta 98), pack-reused 5669
Receiving objects: 100% (5871/5871), 4.12 MiB | 3.81 MiB/s, done.
Resolving deltas: 100% (4073/4073), done.
(dmd-2.099.0)[root@nethserver-test ~]# cd onedrive
(dmd-2.099.0)[root@nethserver-test onedrive]# ./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for dmd... dmd
checking version of D compiler... 2.099.0
checking for curl... yes
checking for sqlite... yes
configure: creating ./config.status
config.status: creating Makefile
config.status: creating contrib/pacman/PKGBUILD
config.status: creating contrib/spec/onedrive.spec
config.status: creating onedrive.1
config.status: creating contrib/systemd/onedrive.service
config.status: creating contrib/systemd/onedrive@.service
(dmd-2.099.0)[root@nethserver-test onedrive]# make clean; make;
rm -f onedrive onedrive.o version
rm -rf autom4te.cache
rm -f config.log config.status
if [ -f .git/HEAD ] ; then \
        git describe --tags > version ; \
else \
        echo v2.4.20-dev > version ; \
fi
dmd  -w -g -O -J. -L-lcurl -L-lsqlite3  -L-ldl src/config.d src/itemdb.d src/log.d src/main.d src/monitor.d src/onedrive.d src/qxor.d src/selective.d src/sqlite.d src/sync.d src/upload.d src/util.d src/progress.d src/arsd/cgi.d -ofonedrive
(dmd-2.099.0)[root@nethserver-test onedrive]# 
(dmd-2.099.0)[root@nethserver-test onedrive]# make install
/usr/bin/install -c -D onedrive /usr/local/bin/onedrive
/usr/bin/install -c -D -m 644 onedrive.1 /usr/local/share/man/man1/onedrive.1
/usr/bin/install -c -D -m 644 contrib/logrotate/onedrive.logrotate /usr/local/etc/logrotate.d/onedrive
mkdir -p /usr/local/share/doc/onedrive
/usr/bin/install -c -D -m 644 README.md config LICENSE CHANGELOG.md docs/Docker.md docs/INSTALL.md docs/SharePoint-Shared-Libraries.md docs/USAGE.md docs/BusinessSharedFolders.md docs/advanced-usage.md /usr/local/share/doc/onedrive
/usr/bin/install -c -d -m 0755 /usr/lib/systemd/user /usr/lib/systemd/system
/usr/bin/install -c -m 0644 contrib/systemd/onedrive@.service /usr/lib/systemd/system
/usr/bin/install -c -m 0644 contrib/systemd/onedrive.service /usr/lib/systemd/system
(dmd-2.099.0)[root@nethserver-test onedrive]# 
(dmd-2.099.0)[root@nethserver-test onedrive]# deactivate
[root@nethserver-test onedrive]# 
[root@nethserver-test onedrive]# cd ~
[root@nethserver-test ~]# 
[root@nethserver-test ~]# onedrive 
Configuring Global Azure AD Endpoints
Authorize this app visiting:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=d50ca740-c83f-4d1b-b616-12c519384f0c&scope=Files.ReadWrite%20Files.ReadWrite.all%20Sites.Read.All%20Sites.ReadWrite.All%20offline_access&response_type=code&prompt=login&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient

Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=XXXXXXX

Application has been successfully authorised, however no additional command switches were provided.

Please use 'onedrive --help' for further assistance in regards to running this application.

[root@nethserver-test ~]# 
```

**systemd setup**
```
Last login: Sun Jul  3 18:07:24 2022 from 172.17.18.11

************ Welcome to NethServer ************

This is a NethServer installation. 

Before editing configuration files, be aware 
of the automatic events and templates system.


          http://docs.nethserver.org

***********************************************
[root@nethserver-test ~]# systemctl enable onedrive
Created symlink from /etc/systemd/system/default.target.wants/onedrive.service to /usr/lib/systemd/system/onedrive.service.
[root@nethserver-test ~]# systemctl start onedrive
[root@nethserver-test ~]# journalctl --unit=onedrive -f
-- Logs begin at Sun 2022-07-03 17:53:17 AEST. --
Jul 03 18:28:04 nethserver-test.local.net onedrive[12401]: Creating local directory: random_files/PavFYfzxOHTEmsUNRrWlVW2LjrWxLPUM
Jul 03 18:28:04 nethserver-test.local.net onedrive[12401]: Creating local directory: random_files/IWy6gsb5gzYfcbRhRCxH9FhwczTXOdE0
Jul 03 18:28:04 nethserver-test.local.net onedrive[12401]: Creating local directory: random_files/icFuaAwgtn1dUpjnMHMFyw9oC37Cajn2
Jul 03 18:28:04 nethserver-test.local.net onedrive[12401]: Creating local directory: random_videos/ZXFgErZFlsS1XuyaaoUplWJeAWWci1xH
Jul 03 18:28:04 nethserver-test.local.net onedrive[12401]: Creating local directory: random_videos/vyhtrZU0BlnBRG4HPs8fUim3ccJzthvf
Jul 03 18:28:04 nethserver-test.local.net onedrive[12401]: Creating local directory: random_videos/kQlCi2oaY6oTxsyRPJSLNT3K0q5KIvq9
Jul 03 18:28:04 nethserver-test.local.net onedrive[12401]: Creating local directory: random_videos/jFq6AIixiNPFF0aYdpMAK7QiaRdggje7
Jul 03 18:28:04 nethserver-test.local.net onedrive[12401]: Creating local directory: random_videos/jl2nmcaiviWQPx3Lswh3pSxJokGMh2Bx
Jul 03 18:28:05 nethserver-test.local.net onedrive[12401]: Downloading file omv_test_files/16.06.19-16.07.03[M][0@0][0].mp4 ... done.
Jul 03 18:28:06 nethserver-test.local.net onedrive[12401]: Downloading file Documents/Images/image001.jpg ... done.
Jul 03 18:28:07 nethserver-test.local.net onedrive[12401]: Downloading file random_files/YOLAgshqE6H1te80EvaRaOxyCGEmlIWB/file0.data ... done.
Jul 03 18:28:08 nethserver-test.local.net onedrive[12401]: Downloading file Documents/Music/audio001.mp3 ... done.
^C
[root@nethserver-test ~]# ps -aufx | grep onedrive 
root     12454  0.0  0.0 112812   984 pts/1    S+   18:28   0:00          \_ grep --color=auto onedrive
root     12401  9.4  0.3 190396 15968 ?        Ss   18:27   0:03 /usr/local/bin/onedrive --monitor
[root@nethserver-test ~]# 
```


I suggest that you strip things back to basics first ... and work from there.

**Comment by abraunegg** [comment]
@tdcockers 
I am flagging this as a bug with the client for the following reason:

**root systemd in debug mode**
```
Jul 03 20:58:45 nethserver-test.local.net systemd[1]: Started OneDrive Free Client.
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] homePath: WARNING - no HOME environment variable set
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] homePath: WARNING - no SHELL environment variable set
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] homePath: WARNING - no USER environment variable set
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] homePath: /root
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] configDirBase: WARNING - no XDG_CONFIG_HOME environment variable set
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] configDirBase: /root/.config
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] configDirName: Configuring application to use default config path
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: Using 'user' Config Dir: /root/.config/onedrive
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: Using 'system' Config Dir: /etc/onedrive
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] refreshTokenFilePath = /root/.config/onedrive/refresh_token
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] deltaLinkFilePath = /root/.config/onedrive/delta_link
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] databaseFilePath = /root/.config/onedrive/items.sqlite3
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] databaseFilePathDryRun = /root/.config/onedrive/items-dryrun.sqlite3
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] uploadStateFilePath = /root/.config/onedrive/resume_upload
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] userConfigFilePath = /root/.config/onedrive/config
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] syncListFilePath = /root/.config/onedrive/sync_list
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] systemConfigFilePath = /etc/onedrive/config
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: [DEBUG] businessSharedFolderFilePath = /root/.config/onedrive/business_shared_folders
Jul 03 20:58:45 nethserver-test.local.net onedrive[24903]: No user or system config file found, using application defaults
```


**custom systemd in debug mode**
```
Jul 03 21:06:45 nethserver-test.local.net systemd[1]: Started OneDrive Free Client.
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] homePath: WARNING - no HOME environment variable set
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] homePath: WARNING - no SHELL environment variable set
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] homePath: WARNING - no USER environment variable set
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] homePath: /root
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] configDirName: CLI override to set configDirName to: "/etc/onedrive/scans"
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: Using 'user' Config Dir: "/etc/onedrive/scans"
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] refreshTokenFilePath = "/etc/onedrive/scans"/refresh_token
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] deltaLinkFilePath = "/etc/onedrive/scans"/delta_link
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] databaseFilePath = "/etc/onedrive/scans"/items.sqlite3
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] databaseFilePathDryRun = "/etc/onedrive/scans"/items-dryrun.sqlite3
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] uploadStateFilePath = "/etc/onedrive/scans"/resume_upload
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] userConfigFilePath = "/etc/onedrive/scans"/config
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] syncListFilePath = "/etc/onedrive/scans"/sync_list
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] systemConfigFilePath = /config
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: [DEBUG] businessSharedFolderFilePath = "/etc/onedrive/scans"/business_shared_folders
Jul 03 21:06:45 nethserver-test.local.net onedrive[25497]: No user or system config file found, using application defaults
```

The difference here is that for whatever reason, extra `"` are being added to the filepath when it is being specified ... which should not be the case.

As such, please can you test the following PR:

```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/2029/head:pr2029
git checkout pr2029
./configure; make clean; make;
```
To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

To install the PR, you will need to perform `sudo make install` to install the PR version. 

When running the PR, your version should be: `onedrive v2.4.19-8-ge81b159` or greater.

**Comment by tdcockers** [comment]
I literally just figured this out 45 seconds ago!! Will test the PR momentarily.

**Comment by tdcockers** [comment]
Hey mate... I believe the PR you've just pushed through is working perfectly, I'm syncing 4 different sharepoint libraries with no issues at all, and all 4 services started on boot.

Thanks for persevering with my edge case to find a resolution despite your initial assessment being an environment specific issue.

**Comment by abraunegg** [comment]
@tdcockers 
> Hey mate... I believe the PR you've just pushed through is working perfectly, I'm syncing 4 different sharepoint libraries with no issues at all, and all 4 services started on boot.

Glad this is now sorted and you are operational, the PR is merged into 'master', also thanks for the sponsorship.

> Thanks for persevering with my edge case to find a resolution despite your initial assessment being an environment specific issue.

Without seeing any logs or evidence of what is going on, and with this particular issue - with something that is known to work - esp if using user or root home directory (which you indicated you were initially using as per above - `/root/.config/onedrive/sharepoint/{various}).` something odd was occurring which was not normal.

It was not until later on where your posts switch to using `/etc/.od-config/scans` or `/etc/onedrive/scans` that I thought it would be prudent to also test / validate this combination on NethServer - given that the normal systemd configuration using the root or normal account was working (`/root/.config/onedrive-custom` or `/home/alex/.config/onedrive-custom/`). When running each of the systemd services using `--verbose --verbose` and watching the systemd logs, this is how this difference and bug was highlighted.
